### PR TITLE
Rewards

### PR DESCRIPTION
**Pull request by 0xJem**
## Summary by CodeRabbit

* **New Features**
  * Epoch-based Merkle incentive distribution for convertible OHM rewards with previewable and multi-epoch claims; deployable convertible OHM option tokens with on-chain minting and exercise; EIP‑2612 permit support for tokens; governance proposal and deployment flows to enable the system; mint cap controls and remaining-approval visibility.

* **Documentation**
  * Added audit scope, architecture docs, and updated role/permission definitions for the new incentive system.

* **Tests**
  * Extensive new test suites covering deployment, endEpoch, claim/preview, mint/create, exercise, admin controls, and end-to-end flows.

### CODE REVIEW

**Comment by 0xJem** [comment]
@0xNooodle can you resolve the conflicts?

File: src/policies/incentives/convertible/ConvertibleOHMTeller.sol [hunk 1]
```diff
@@ -61,12 +63,16 @@ contract ConvertibleOHMTeller is
...
-    uint256 private constant _ENABLE_DATA_LENGTH = 32;
+    /// @notice Minimum byte length of enableData_ (one ABI-encoded uint256 mintCap + two dynamic arrays)
+    /// @dev abi.encode(uint256, address[], uint256[]) with empty arrays = 32 + 32 + 32 + 32 + 32 = 160
+    uint256 private constant _MIN_ENABLE_DATA_LENGTH = 160;
 
     /// @notice Minimum decimals required for quote tokens (used by _formatPrice)
     uint8 private constant _MIN_QUOTE_TOKEN_DECIMALS = 2;
 
+    /// @notice Maximum allowed duration from current time to token expiry (~2 years)
+    uint48 private constant _MAX_EXPIRY_HORIZON = 730 days;
```
Comment (0xJem) [inline src/policies/incentives/convertible/ConvertibleOHMTeller.sol:83]: why this? It feels arbitrary

File: src/policies/incentives/convertible/ConvertibleOHMTeller.sol [hunk 2]
```diff
@@ -172,17 +200,24 @@ contract ConvertibleOHMTeller is
         nonReentrant
         returns (address)
     {
-        // If eligible is zero, use the current timestamp
-        if (eligible_ == 0) eligible_ = uint48(block.timestamp);
+        if (eligible_ == 0) {
```
Comment (0xJem) [inline src/policies/incentives/convertible/ConvertibleOHMTeller.sol:237]: This feels off - needlessly complex. The `1 days - 1` also feels arbitrary. If `1 days` is the minimum for eligibility, then make it a constant. I also think there need to be fuzz tests for different values of `minEligibleDelay`, as well as boundary tests (e.g. 1 day)

File: src/policies/incentives/convertible/ConvertibleOHMTeller.sol [hunk 3]
```diff
@@ -172,17 +200,24 @@ contract ConvertibleOHMTeller is
...
+            eligible_ = _truncateToUTCDay(minTimestamp + 1 days - 1);
+        }
 
         // Note: Convertible tokens are only unique to a day, not a specific timestamp
         (eligible_, expiry_) = _truncateBothToUTCDay(eligible_, expiry_);
 
         // Revert if eligible is in the past, we do this to avoid duplicates tokens with the same parameters otherwise
         // Truncate block.timestamp to the nearest day for comparison
-        if (eligible_ < _truncateToUTCDay(uint48(block.timestamp)))
+        if (eligible_ < uint48(block.timestamp) + minEligibleDelay)
```
Comment (0xJem) [inline src/policies/incentives/convertible/ConvertibleOHMTeller.sol:248]: Need to update the comment

File: src/policies/incentives/convertible/ConvertibleOHMTeller.sol (hunk 2 above)
Comment (0xJem) [reply src/policies/incentives/convertible/ConvertibleOHMTeller.sol:237]: If this is meant to fix TRSRT-L-2, then add a specific test for eligibility in the past

File: src/test/policies/incentives/ConvertibleOHMTeller.t.sol [hunk 4]
```diff
@@ -1257,6 +1495,51 @@ contract ConvertibleOHMTellerAdminTests is ConvertibleOHMTellerTestBase {
         );
     }
 
+    function test_enable_withCreatorLimits() external {
```
Comment (0xJem) [inline src/test/policies/incentives/ConvertibleOHMTeller.t.sol]: fuzz limits
add a test for multiple creators and limits

File: src/test/policies/incentives/ConvertibleOHMTeller.t.sol [hunk 5]
```diff
@@ -1110,6 +1215,137 @@ contract ConvertibleOHMTellerAdminTests is ConvertibleOHMTellerTestBase {
...
+    // ========== exercise: mint limit ========== //
+
+    function test_exercise_revertsIfMintLimitExceeded() external {
+        // 1. Preparation: set a low mint limit for the creator
+        teller.setAdminMintLimit(incentiveDistributor, 50e9);
+
+        // Deploy and mint
+        ConvertibleOHMToken token = _deployConvertibleToken();
+        vm.prank(incentiveDistributor);
+        teller.create(address(token), user0, 100e9);
```
Comment (0xJem) [inline src/test/policies/incentives/ConvertibleOHMTeller.t.sol:1552]: Design Q: why doesn't it revert at the time `create()` is called? It feels unfair to the recipient that they can receive the convOHM token, but are unable to mint it.

File: src/policies/incentives/convertible/ConvertibleOHMTeller.sol
Comment (0xJem) [inline src/policies/incentives/convertible/ConvertibleOHMTeller.sol]: This still allows the teller admin to set the mint cap, which in turn affects the MINTR approval. It's also a little confusing to have the admin mint limit + mint cap (and they could get out of sync).

Why not simply have a per-creator mint cap (accessible only by the admin role)?
This would tie together token creation + exercise limits.

Example:
- set creatorA to have a mint cap of 50e9, which increases the MINTR approval for the teller by 50e9
- creating 10e9 convOHM tokens by `creatorA` (10e9 < 50e9 mint cap, so OK)
- creating 40e9+1 convOHM tokens by `creatorA` (10e9 + 40e9 + 1 > 50e9 mint cap, so revert)
- exercise and convert 10e9 to OHM (MINTR approval decreases by 10e9)
- need to figure out how to handle changing the mint cap for creatorA later (old mint cap, new mint cap, exercised options)

File: src/policies/incentives/convertible/ConvertibleOHMTeller.sol [hunk 6]
```diff
@@ -551,7 +598,37 @@ contract ConvertibleOHMTeller is
...
+    }
+
+    function _setMinDuration(uint48 duration_) private {
         minDuration = duration_;
+        emit MinDurationSet(duration_);
+    }
+
+    /// @inheritdoc IConvertibleOHMTeller
+    function setMinEligibleDelay(uint48 delay_) external onlyEnabled onlyAdminRole {
+        if (delay_ < uint48(1 days)) revert Teller_InvalidParams(0, abi.encodePacked(delay_));
```
Comment (0xJem) [inline src/policies/incentives/convertible/ConvertibleOHMTeller.sol]: extract magic number into constant

File: src/policies/incentives/convertible/ConvertibleOHMTeller.sol (hunk 1 above)
Comment (zeroxnoodle) [reply src/policies/incentives/convertible/ConvertibleOHMTeller.sol:83]: TRST-R-5 recommendation

File: src/policies/incentives/convertible/ConvertibleOHMTeller.sol (hunk 6 above)
Comment (Yurii3721) [reply src/policies/incentives/convertible/ConvertibleOHMTeller.sol]: Done: https://github.com/OlympusDAO/olympus-v3/pull/233

File: src/test/policies/incentives/ConvertibleOHMTeller.t.sol (hunk 4 above)
Comment (Yurii3721) [reply src/test/policies/incentives/ConvertibleOHMTeller.t.sol]: Done: https://github.com/OlympusDAO/olympus-v3/pull/234

File: src/policies/incentives/convertible/ConvertibleOHMTeller.sol (hunk 3 above)
Comment (Yurii3721) [reply src/policies/incentives/convertible/ConvertibleOHMTeller.sol:248]: Done: https://github.com/OlympusDAO/olympus-v3/pull/233

File: src/policies/incentives/convertible/ConvertibleOHMTeller.sol (hunk 2 above)
Comment (Yurii3721) [reply src/policies/incentives/convertible/ConvertibleOHMTeller.sol:237]: Added the tests (https://github.com/OlympusDAO/olympus-v3/pull/234) and named the constants (https://github.com/OlympusDAO/olympus-v3/pull/233).

This is the TRST‑L‑2 fix that the auditors suggested in their report.

The expression `_truncateToUTCDay(minTimestamp + _UTC_DAY - 1)` acts as a ceiling to the next UTC midnight (i.e., `(x + divisor – 1) / divisor * divisor`).

After truncation, `eligible_` should be at UTC midnight because tokens are identified by day. A plain `minTimestamp` almost never lands exactly on midnight, and `_truncateToUTCDay(minTimestamp)` would round down, producing a value **earlier** than `minTimestamp` and consequently failing the check `eligible_ < block.timestamp + minEligibleDelay`.

Simplifying this without breaking correctness cloud be difficult, ceiling‑to‑midnight is a simple solution.

File: src/test/policies/incentives/ConvertibleOHMTeller.t.sol (hunk 5 above)
Comment (Yurii3721) [reply src/test/policies/incentives/ConvertibleOHMTeller.t.sol:1552]: Review: https://github.com/OlympusDAO/olympus-v3/pull/235

File: src/policies/incentives/convertible/ConvertibleOHMTeller.sol
Comment (Yurii3721) [reply src/policies/incentives/convertible/ConvertibleOHMTeller.sol]: Review https://github.com/OlympusDAO/olympus-v3/pull/235 before updating scripts and tests.

File: src/policies/incentives/convertible/interfaces/IConvertibleOHMTeller.sol [hunk 7]
```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: AGPL-3.0
+pragma solidity >=0.8.4;
```
[the remark this answers is not in the document]
Comment (0xJem) [reply src/policies/incentives/convertible/interfaces/IConvertibleOHMTeller.sol:2]: @zeroxnoodle @Yurii3721 can you make these changes? MIT, `>=0.8.15`

File: src/proposals/IncentiveDistributorProposalConvertible.sol [hunk 8]
```diff
@@ -2,21 +2,24 @@
 /// forge-lint: disable-start(mixed-case-function, mixed-case-variable)
 // solhint-disable one-contract-per-file
 // solhint-disable custom-errors
-pragma solidity >=0.8.20;
+pragma solidity >=0.8.30;
```
Comment (0xJem) [inline src/proposals/IncentiveDistributorProposalConvertible.sol:5]: why 0.8.30? I'd use `>=0.8.24` unless there's a good reason

File: src/policies/incentives/IncentiveDistributorConvertible.sol [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    /// @inheritdoc IIncentiveDistributorConvertible
+    function previewClaim(
+        address user_,
+        uint256[] calldata epochEndDates_,
+        uint256[] calldata amounts_,
+        bytes32[][] calldata proofs_
+    ) external view returns (address[] memory tokens, uint256[] memory claimableAmounts) {
+        uint256 len = epochEndDates_.length;
+        if (len == 0 || len != amounts_.length || len != proofs_.length)
+            return (tokens, claimableAmounts);
```
[the remark this answers is not in the document]
Comment (0xJem) [reply src/policies/incentives/IncentiveDistributorConvertible.sol:167]: @zeroxnoodle take a look at this

File: src/policies/incentives/BaseIncentiveDistributor.sol [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+        if (epochMerkleRoots[epochEndDate_] != bytes32(0))
+            revert IncentiveDistributor_EpochAlreadySet(epochEndDate_);
+
+        // Validate epochEndDate is at 23:59:59 UTC (end of day)
+        _validateEpochEndOfDay(epochEndDate_);
+
+        // Validate epochEndDate is at least 1 day after lastEpochEndDate
+        if (epochEndDate_ < lastEpochEndDate + MIN_EPOCH_DURATION) {
+            revert IncentiveDistributor_EpochTooEarly();
+        }
```
[the remark this answers is not in the document]
Comment (0xJem) [reply src/policies/incentives/BaseIncentiveDistributor.sol:105]: @zeroxnoodle valid?

File: src/proposals/IncentiveDistributorProposalConvertible.sol [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+                "and exercise them for OHM by paying the conversion price in the quote token.\n\n",
+                "The ConvertibleOHMTeller is deployed with the following defaults (adjustable by governance):\n",
+                "- `minDuration` = 1 day: minimum exercise window between eligible and expiry.\n",
+                "- `minEligibleDelay` = 1 day: minimum delay between token deployment and eligibility, ",
+                "providing the emergency role a time window to disable the contract if needed.\n\n",
+                "The per-creator mint cap limits the cumulative convOHM the distributor can ever mint. ",
+                "MINTR approval is adjusted atomically on create / exercise / sweep. ",
+                "Expired tokens free up MINTR approval periodically when Heart sweeps them.\n\n",
+                "## References\n\n",
+                "TODO: Add RFC/OIP reference.\n",
```
Comment (0xJem) [inline src/proposals/IncentiveDistributorProposalConvertible.sol:90]: @zeroxnoodle some TODOs here

Audit report: https://storage.googleapis.com/olympusdao-landing-page-reports/audits/2026-04_Engage.pdf

File: src/scripts/deploy/savedDeployments/convertible_incentive_distributor.json [hunk 12]
```diff
@@ -0,0 +1,14 @@
+{
+    "sequence": [
+        {
+            "name": "ConvertibleOHMTeller",
+            "args": {}
+        },
+        {
+            "name": "IncentiveDistributorConvertible",
+            "args": {
+                "lastEpochEndDate": 1753919999
```
Comment (0xJem) [inline src/scripts/deploy/savedDeployments/convertible_incentive_distributor.json:10]: This needs to be updated

File: src/proposals/IncentiveDistributorProposalConvertible.sol [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+/// @notice Proposal to enable the Convertible OHM Incentive Distributor system
+contract IncentiveDistributorProposalConvertible is GovernorBravoProposal {
+    Kernel internal _kernel;
+
+    // ========== CONSTANTS ========== //
+
+    /// TODO: Decide on the initial distributor mint cap
+    /// @notice Per-creator mint cap (max outstanding convOHM) for the IncentiveDistributorConvertible (in OHM units, 9 decimals).
+    ///         Also determines the initial MINTR approval head-room available to the teller.
+    uint256 internal constant _INITIAL_DISTRIBUTOR_MINT_CAP = 1000e9;
```
Comment (0xJem) [inline src/proposals/IncentiveDistributorProposalConvertible.sol:34]: Set to 10,000

File: src/proposals/IncentiveDistributorProposalConvertible.sol [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+    /// forge-lint: disable-next-line(unsafe-typecast)
+    bytes32 internal constant _ROLE_CONVERTIBLE_DISTRIBUTOR = bytes32("convertible_distributor");
+    /// forge-lint: disable-next-line(unsafe-typecast)
+    bytes32 internal constant _ROLE_INCENTIVE_MANAGER = bytes32("incentive_manager");
+
+    /// @notice Expected periodic task count on the Heart after this proposal is executed.
+    ///         The teller is appended at the end, so the count is the existing pre-proposal
+    ///         count plus one. Snapshotted here to catch unexpected drift in the Heart's
+    ///         periodic-task list between proposal authoring and execution.
+    uint256 internal constant _EXPECTED_PERIODIC_TASK_COUNT = 7;
```
Comment (0xJem) [inline src/proposals/IncentiveDistributorProposalConvertible.sol:45]: Check indexes
